### PR TITLE
Restructure how resource allocation/deallocation/read steps are defined and executed

### PR DESCRIPTION
**Pull request by tasdomas**
Defining steps as a slice of structures at least removes the necessity
to renumber the progress indicator when inserting new steps (or removin
them).

### CODE REVIEW

**Comment by 0x2b3bfa0** [comment]
Related to https://github.com/iterative/terraform-provider-iterative/issues/290; awesome enhancement!

**Review by 0x2b3bfa0 — CHANGES_REQUESTED** [review]
AWS tests are [failing](https://github.com/iterative/terraform-provider-iterative/runs/7881775867?check_suite_focus=true#step:5:217)

**Comment by tasdomas** [comment]
> AWS tests are [failing](https://github.com/iterative/terraform-provider-iterative/runs/7881775867?check_suite_focus=true#step:5:217)

yes, just saw that - will take care of it tomorrow

**Comment by tasdomas** [comment]
> AWS tests are [failing](https://github.com/iterative/terraform-provider-iterative/runs/7881775867?check_suite_focus=true#step:5:217)

All fixed now!

**Review by dacbd — APPROVED** [review]
lgtm

**Review by 0x2b3bfa0 — CHANGES_REQUESTED** [review]
I'd like `k8s` smoke test to pass at least once before merging, it seems to be failing because of unknown reasons not related to resource availability, and this pull request alters lots of fragile code.

**Comment by dacbd** [comment]
I will give the K8s tests a couple of re-runs late PST time, /reasons 🔌 🌎 ⏳

**Comment by tasdomas** [comment]
> I'd like `k8s` smoke test to pass at least once before merging, it seems to be failing because of unknown reasons not related to resource availability, and this pull request alters lots of fragile code.

The k8s test failures were a genuine issue and not smoke test flakiness - it's fixed now

File: task/k8s/task.go [hunk 1]
```diff
@@ -184,44 +183,42 @@ func (t *Task) Read(ctx context.Context) error {
...
+			Description: "Creating ephemeral Job to retrieve directory...",
+			Action:      withEnv(env, t.Resources.Job.Create),
+		}, {
+			Description: "Downloading Directory...",
+			Action: withEnv(env, func(ctx context.Context) error {
+				return t.Pull(ctx, t.Attributes.Directory, t.Attributes.DirectoryOut)
+			}),
+		}, {
+			// WTH?
+			Description: "Deleting ephemeral Job to retrieve directory...",
```
Comment (tasdomas) [inline task/k8s/task.go:206]: Smoke tests fail if we replace `t.Resources.Job.Create` with `Delete`

File: task/k8s/task.go [hunk 2]
```diff
@@ -302,3 +299,14 @@ func (t *Task) GetKeyPair(ctx context.Context) (*ssh.DeterministicSSHKeyPair, er
 func (t *Task) GetIdentifier(ctx context.Context) common.Identifier {
 	return t.Identifier
 }
+
+// withEnv runs the specified action with the environment variables set.
+func withEnv(env map[string]string, action func(context.Context) error) func(context.Context) error {
```
Comment (tasdomas) [inline task/k8s/task.go:304]: This is only needed because we de-facto use env vars instead of passing function params. This needs to be fixed ASAP

**Review by 0x2b3bfa0 — APPROVED** [review]

File: task/k8s/task.go (hunk 1 above)
Comment (0x2b3bfa0) [reply task/k8s/task.go:206]: Resolving in favor of https://github.com/iterative/terraform-provider-iterative/issues/647.

File: task/k8s/task.go (hunk 2 above)
Comment (0x2b3bfa0) [reply task/k8s/task.go:304]: Resolving in favor of https://github.com/iterative/terraform-provider-iterative/issues/648.

**Pull request merged**